spices/SPICE-0009-external-readers.adoc

```diff
@@ -0,0 +1,335 @@
+This hinders development and debugging of such modules, as typical workflows involving `trace(<expr>)` or `pkl eval -x '<expr>'` behave differently or are no longer possible.
+
+== Proposed Solution
+
+The existing message passing API and client libraries will be expanded to support external readers.
+Instead of client libraries invoking `pkl server`, `pkl eval` invocations, when configured with an external reader, will run the reader executable as a subprocess.
+External readers can provide any number of resource and/or module schemes and will be configured as an executable and list of arguments via:
+
+* CLI flag
+* PklProject property
+* Java API
+* Message passing API
+
+[source,pkl]
+----
+/// May be specified as an absolute path to an executable
+/// May also be specified as just an executable name, in which case it will be resolved according to the PATH environment variable
+executable: String
+
+/// Command line arguments that will be passed to the reader process
+arguments: Listing<String>
+----
+
+=== Example
+
+Consider this module:
+
+[source,pkl]
+----
+username = "john_appleseed"
+
+email = read("ldap://ds.example.com:389/dc=example,dc=com?mail?sub?(uid=\(username))").text
+----
+
+Pkl doesn't implement the `ldap:` resource URI scheme natively, but an external reader can provide it.
+Assuming a hypothetical `pkl-ldap` binary implementing the external reader protocol and the `ldap:` scheme is in the `$PATH`, this module can be evaluated as:
+
+[source,text]
+----
+$ pkl eval <module> --external-reader=pkl-ldap --allowed-resources prop:,ldap:
+username = "john_appleseed"
+email = "appleseed@example.com"
+----
+
+For security reasons, schemes provided by external readers are not allowed by default and clients must explicitly allow the desired scheme(s).
+In this example, the external reader may provide both `ldap:` and `ldaps:` schemes.
+If a codebase or user wishes to enforce that only secure LDAP connections are made, the `ldaps:` scheme may be allowed while `ldap:` is not.
+
+== Detailed design
+
+To avoid terminology confusion with the existing language binding message passing model, the `pkl` process will remain known as the "server" and the reader process will be known as the "client", despite all requests in this proposal being initiated by the `pkl` process.
+
+=== Overall Flow
+
+1. User requests an evaluator (via Java API, Message Passing API, or CLI) with an external reader.
+2. Entrypoint (client Java code, `CliEvaluator`, or `Server`) instantiates an `ExternalReader`.
+3. The `ExternalReader` is started, which starts the configured child process and sends a `DiscoverReadersRequest` message.
+4. A `DiscoverReadersResponse` message is received (within some timeout).
```

Comment:
Thinking about this a little more--what's the use-case for wanting discovery?

Asides from our `GenericUrl` catch-all reader in Java, I'm not really aware of other use-cases for having module readers handle unknown schemes ahead of time. The rest of our readers handle a statically known set of schemes.

The API of our message passing API is more limited, and requires that each reader's scheme be known ahead of time. We did this because it wasn't really clear if we ever needed a more "generic" reader model that the Java side has.

Reply:
Another benefit of requiring users to declare schemes ahead of time is that we don't need to spawn these external processes unless executing an `import` or `read` that uses their scheme. This can be important if, say, we allow external readers to be configured globally per machine (e.g. by adding them to `~/.pkl/settings.pkl`).

Reply:
The biggest motivation is user experience.

Here are three takes on discovery:

## Full discovery (as proposed)

The supported schemes (module and resource) and their local/globbable/hierarchical flags can be discovered purely with knowledge on how to invoke the process. This yields the most "on rails" user experience:

```console
--external-reader pkl-ldap
# or
--external-reader pkl-myproto
```
Or in PklProject
```pkl
evaluatorSettings {
  externalReaders {
    new { executable = "pkl-ldap" }
  }
}
// or
evaluatorSettings {
  externalReaders {
    new { executable = "pkl-myproto" }
  }
}
```

## No discovery at all

In a world without any discovery, the scheme mapping and local/globbable/hierarchical flags must be configured up front. The config must look something like this:
```console
--external-resource-reader ldap=pkl-ldap --external-resource ldaps=pkl-ldap
# or
--external-resource-reader myproto=pkl-myproto --external-resource-globabable myproto --external-resource-hierarchical myproto --external-resource-local myproto --external-module-reader myproto=pkl-myproto --external-module-globabable myproto --external-module-hierarchical myproto --external-module-local myproto
```
Or in PklProject
```pkl
evaluatorSettings {
  externalResourceReaders {
    new { scheme = "ldap"; executable = "pkl-ldap" }
    new { scheme = "ldaps"; executable = "pkl-ldap" }
  }
}
// or
evaluatorSettings {
  externalResourceReaders {
    new { scheme = "myproto"; executable = "pkl-myproto"; globbable = true; hierarchical = true; local = true }
  }
  externalModuleReaders {
    new { scheme = "myproto"; executable = "pkl-myproto"; globbable = true; hierarchical = true; local = true }
  }
}
```

This is a ton of config for users to get wrong. Reader implementations already know these properties about themselves and should be able to answer for them.

## Minimal discover

A compromise here is to require manual scheme mapping still, but launch the process on first read and perform discovery of a reader process's local/globbable/hierarchical flags for that module/resource scheme. Config might look like this;
```console
--external-resource-reader ldap=pkl-ldap --external-resource-reader ldaps=pkl-ldap
# or
--external-resource-reader myproto=pkl-myproto --external-module-reader myproto=pkl-myproto
```
Or in PklProject
```pkl
evaluatorSettings {
  externalResourceReaders {
    new { scheme = "ldap"; executable = "pkl-ldap" }
    new { scheme = "ldaps"; executable = "pkl-ldap" }
  }
}
// or
evaluatorSettings {
  externalResourceReaders {
    new { scheme = "myproto"; executable = "pkl-myproto" }
  }
  externalModuleReaders {
    new { scheme = "myproto"; executable = "pkl-myproto" }
  }
}
```

I still don't think this is ideal from a user experience perspective, but it does allow very straightforward deferral of process spawn until first read.

Reply:
I think I've successfully convinced myself that the minimal discovery mechanism has enough benefits with few enough downsides to be the correct path forward here. I'm working on revising the SPICE to reflect that.

This choice raises some more questions:
* Should explicitly registering a reader for a scheme automatically allow resource/module reads for that scheme?
  * This would be ignored if explicit allow-lists are configured, which matches the behavior for built-in readers
  * This is a nice tradeoff for the slightly more involved configuration in simple cases
* Should any kind of warning be produced when built-in readers are overridden?
  * I can see a lot of potential for nasty surprises when doing this, especially if it becomes possible to configure this via `~/.pkl/settings.pkl`.

Reply:
Agree with Dan here that having a statically defined mapping of external readers is better. Not only is much more performant (Waiting on multiple readers for a scheme listing response may tank performance, specially for small configs), it's also more readable: you can see directly in the cli/setting.pkl/PklProject exactly which schemes are supported, no needed to look into the readers themselves.

I disagree that making it explicit makes the configuration more complicated. The user has to know which schemes the reader supports anyway, because they have to write the read/import string with that scheme.

Another benefit: let's say a reader supports three schemes but I just want to use one in my project. I'd just add that one and I don't have to worry about other schemes sneaking in into my code, making it safer.

Reply:
I think the "minimal" approach makes sense. Flags like hierarchical/globbable/local are implementation details of a reader, so it's kind of nice that users don't need to configure it explicitly. However, the scheme is an important aspect of the reader that both implementers and users should be aware of, so it makes sense to me that they need to provide it to the evaluator.

> Should explicitly registering a reader for a scheme automatically allow resource/module reads for that scheme?

I think it depends; for the CLI, `pkl:Project` and `pkl:settings`, it might make to add its scheme to the default set of allowed resources/modules if not otherwise configured:

But I don't think this makes sense for `CreateEvaluatorRequest`, nor for the evaluator builder. Those are lower level, and users can build helpers easily to configure these.

We can also add helpers in pkl-swift and pkl-go to set them. We already have this today with adding module/resource readers.

> Should any kind of warning be produced when built-in readers are overridden?

The flip side is that can be quite annoying for users that know what they're doing and intentionally set them. I don't feel strongly about this right now, though.